### ISSUE TITLE
fix: align numpy and JAX Hessian paths in LensCalc via Richardson extrapolation

### ISSUE THREAD

**Issue by Jammy2211** [Jammy2211/PyAutoGalaxy#357]
## Overview

`LensCalc.hessian_from` dispatches to two different implementations depending on
the `xp` backend: the numpy path uses 2-point central finite differences with
`buffer=0.01` (truncation error O(h²) ≈ 1e-4), while the JAX path uses exact
`jax.jacfwd` autodiff. At near-critical points, this ~1e-4 error in the Hessian
gets amplified into ~5e-4 relative errors in derived magnifications, which
produces ~0.1% log-likelihood drift in the point-source full-pipeline JIT
(forcing the regression assertion to a loose `rtol=2e-3`). This task closes
the gap generically with Richardson extrapolation on the existing stencil.

## Plan

- Keep the existing 2-point central-difference stencil as the primitive — no
  mass-profile API changes, no new methods on any profile.
- Wrap it with Richardson extrapolation in the `xp is np` branch: evaluate the
  Hessian at step sizes `h` and `h/2`, combine as `(4·H(h/2) − H(h))/3`. This
  cancels the leading O(h²) error, leaving O(h⁴) ≈ 1e-8 — matching the JAX
  path to float64.
- Cost is 2× the deflection calls (4 → 8 per grid), negligible compared to a
  full non-linear search iteration.
- Add unit tests asserting np vs jnp agreement to `rtol=1e-8` for both the raw
  Hessian tuple and the derived magnification.
- Once merged, tighten the workspace JIT regression assertion in
  `point_source/source_plane.py` from `rtol=2e-3` back to `rtol=1e-4` and drop
  the "pending parity fix" comment.
- Explicitly out of scope: analytic Hessians via profile-specific
  `convergence_2d_from` + `shear_yx_2d_from`. Clean and exact, but would
  require new API surface on every mass profile — noted for future work only.

<details>
<summary>Detailed implementation plan</summary>

### Affected Repositories
- PyAutoGalaxy (primary)
- autolens_workspace_developer (workspace follow-up — tightens regression rtol)

### Work Classification
Both — starts as library work, workspace follow-up lands after shipping the
library fix.

### Branch Survey

| Repository | Current Branch | Dirty? |
|-----------|---------------|--------|
| ./PyAutoGalaxy | main | clean |
| ./autolens_workspace_developer | main | clean |

**Suggested branch:** `feature/lens-calc-hessian-richardson`
**Worktree root:** `~/Code/PyAutoLabs-wt/lens-calc-hessian-richardson/` (created later by `/start_library`)

### Implementation Steps

1. In `PyAutoGalaxy/autogalaxy/operate/lens_calc.py`, leave
   `_hessian_via_finite_difference(self, grid, buffer=0.01)` (line 413)
   unchanged — it stays the primitive stencil.
2. Rewrite the `xp is np` branch of `hessian_from` (line 379) so it calls
   `_hessian_via_finite_difference` twice (at `h=0.01` and `h/2=0.005`) and
   combines each of the four Hessian components with
   `(4·H(h/2) − H(h))/3`. The `xp is jnp` path stays as-is.
3. Update the `hessian_from` docstring to mention that the numpy path is
   Richardson-extrapolated to O(h⁴) for float64 parity with the JAX path.
4. Add unit tests in `test_autogalaxy/operate/test_lens_calc.py`:
   - `test__hessian_from__np_richardson_matches_jax_jacfwd_to_float64` — builds
     a small tracer on a `Grid2DIrregular` and asserts
     `np.allclose(np_hess, jnp_hess, rtol=1e-8)` for all four Hessian
     components.
   - `test__magnification_2d_via_hessian_from__np_jnp_agree_to_float64` — same
     tracer/grid, asserts the same rtol on the derived magnification.
5. Run the full `test_autogalaxy/operate/` test suite to catch any regression
   on existing callers of `hessian_from` (convergence, shear, magnification,
   Jacobian).
6. Workspace follow-up (after library PR merges):
   - Re-run `autolens_workspace_developer/jax_profiling/point_source/source_plane.py`.
   - Tighten the `if full_pipeline_jits:` regression assertion (≈ line 522)
     from `rtol=2e-3` to `rtol=1e-4` against
     `EXPECTED_LOG_LIKELIHOOD_SOURCE_PLANE = -4496.798984131583`.
   - Remove the "pending the np/jnp parity fix" comment that points at this
     task.

### Key Decisions / Trade-offs

- **Richardson vs 5-point stencil** — both give O(h⁴) at 8 deflection calls per
  grid. Richardson wins because it reuses the existing 2-point stencil
  verbatim (trivial wrapper); 5-point requires rewriting the FD arithmetic.
- **Richardson vs analytic Hessians** — analytic is exact (not O(h⁴)), but
  requires every mass profile to expose a matching `convergence_2d_from` /
  `shear_yx_2d_from` pair. New per-profile API surface, large blast radius.
  Deferred as future work.
- **Shrinking the buffer** (e.g. 1e-5) was considered and rejected — at that
  step size, float64 subtraction round-off dominates and the result gets
  worse, not better. Richardson is the clean way to reduce truncation
  without paying round-off costs.

### Key Files
- `PyAutoGalaxy/autogalaxy/operate/lens_calc.py` — rewrite `xp is np` branch of
  `hessian_from` (line 379); add Richardson wrapper around
  `_hessian_via_finite_difference`.
- `test_autogalaxy/operate/test_lens_calc.py` (or nearest equivalent) — add the
  two np/jnp parity tests described above.
- `autolens_workspace_developer/jax_profiling/point_source/source_plane.py` —
  workspace follow-up: tighten `rtol=2e-3` back to `rtol=1e-4` (≈ line 522).

</details>

## Original Prompt

<details>
<summary>Click to expand starting prompt</summary>

# LensCalc.magnification_2d_via_hessian_from diverges between np and jnp

## Context

Follow-up from the `grid-irregular-xp-propagation` task. After fixing
`Grid2DIrregular.grid_2d_via_deflection_grid_from` to propagate `xp` and
rewrapping observed positions + magnifications inside
`AbstractFitPoint`, the full-pipeline source-plane JIT now runs end-to-end —
no more `TracerArrayConversionError`.

However the JIT-path log-likelihood disagrees with the eager-path result by
~0.1%, well outside float64 precision:

- eager (xp=np):  `-4496.798984`
- jax   (xp=jnp): `-4491.832094`  (diff ≈ 4.97, rtol ≈ 1.1e-3)

The `ray_trace_to_source_plane` prefix already asserts equal-to-1e-4 against
the numpy path, and standalone diagnosis confirmed:

- `residual_map` np vs jnp agrees to ~1e-8 (the ray-trace is identical).
- `magnifications_at_positions` is where the numbers diverge.

Concretely, evaluated at the same prior-median parameters on the same two
observed positions:

| Position | np magnification | jnp magnification | rel. diff |
|----------|------------------|-------------------|-----------|
| 0        | 10.60783132      | 10.6070619        | 7e-5      |
| 1        | 152.66911098     | 152.5846727       | 5e-4      |

Because chi-squared divides by `mag**-2 * noise**2`, even a 5e-4 relative
error on one point gets amplified into the ~0.1% log-likelihood drift seen
above.

## Root cause

Traced to `ag.LensCalc.hessian_from` in
`PyAutoGalaxy/autogalaxy/operate/lens_calc.py:350`. It dispatches on `xp`:

- `xp is np` → `_hessian_via_finite_difference` (lens_calc.py:413) — uses
  2-point central difference with `buffer=0.01`. Truncation error is
  **O(h²) ≈ 1e-4**.
- `xp is jnp` → `_hessian_via_jax` (lens_calc.py:383) — uses `jax.jacfwd`
  on `deflections_yx_scalar` vectorised with `jnp.vectorize`. This is
  **exact** forward-mode autodiff — accurate to float64.

So the two paths aren't "both mathematically correct and should match" —
the JAX path is more accurate, and the numpy path is truncation-limited.
At near-critical points (position 1 in the repro above, magnification = 152)
the O(1e-4) truncation error in the Hessian is amplified by the near-zero
Jacobian determinant, producing the observed ~5e-4 relative error in the
magnification.

## Task

Close the accuracy gap on the **numpy** path so it agrees with the JAX
path to float64 precision (~1e-8 rtol). Do this generically, without
touching any mass-profile API.

### Strategy — Richardson extrapolation of the existing FD stencil

Evaluate the existing central-difference Hessian at two step sizes (`h` and
`h/2`) and combine:

```
H_extrapolated = (4 * H(h/2) - H(h)) / 3
```

This cancels the leading O(h²) error term, leaving O(h⁴) ≈ 1e-8 — which
should match the JAX path. It is a pure wrapper around the existing
`_hessian_via_finite_difference(grid, buffer=h)`; no change to stencils,
no change to mass profiles, no new public API.

### Why not the alternatives

- **Analytic Hessians** via profile-specific `convergence_2d_from` +
  `shear_yx_2d_from`. Exact, but requires every mass profile to expose a
  matching pair of methods — new API surface and per-profile work. Noted
  as future work only; **out of scope for this task**.
- **5-point central stencil** — same O(h⁴) accuracy as Richardson at the
  same cost (4 evaluations per axis), but requires rewriting the stencil
  arithmetic in-place. Richardson wins on minimal-churn: the existing
  2-point code stays untouched.
- **Shrinking `buffer`** (e.g. 1e-5) — hits floating-point round-off and
  gets worse, not better.

### Implementation steps

1. In `PyAutoGalaxy/autogalaxy/operate/lens_calc.py`, leave
   `_hessian_via_finite_difference(self, grid, buffer=0.01)` unchanged —
   it remains the primitive stencil.
2. Rewrite the `xp is np` branch of `hessian_from` (lens_calc.py:379) to
   compute Richardson-extrapolated Hessians:

   ```python
   if xp is np:
       h = 0.01
       yy_h,  xy_h,  yx_h,  xx_h  = self._hessian_via_finite_difference(grid, buffer=h)
       yy_h2, xy_h2, yx_h2, xx_h2 = self._hessian_via_finite_difference(grid, buffer=h / 2)
       # Richardson: cancels leading O(h^2) error -> O(h^4).
       yy = (4.0 * yy_h2 - yy_h) / 3.0
       xy = (4.0 * xy_h2 - xy_h) / 3.0
       yx = (4.0 * yx_h2 - yx_h) / 3.0
       xx = (4.0 * xx_h2 - xx_h) / 3.0
       return yy, xy, yx, xx
   return self._hessian_via_jax(grid=grid, xp=xp)
   ```

   (Exact structure is at the implementer's discretion — a private helper
   is fine too.)
3. Verify convergence locally:
   - Pick a mass profile (e.g. `mp.Isothermal`) and a small `Grid2DIrregular`.
   - Compute `np` Hessian with the new Richardson path and `jnp` Hessian with
     the JAX path.
   - Assert `np.allclose(np_hessian, jnp_hessian, rtol=1e-8)`.

### Tests to add

Add a unit test in `test_autogalaxy/operate/test_lens_calc.py` (or the
closest existing file):

- `test__hessian_from__np_richardson_matches_jax_jacfwd_to_float64` — builds
  a small tracer and asserts `np.allclose(np_hess, jnp_hess, rtol=1e-8)` for
  all four Hessian components on the same grid.
- `test__magnification_2d_via_hessian_from__np_jnp_agree_to_float64` — same
  tracer, asserts the same rtol on the derived magnification.

### Workspace follow-up

Once the library fix lands:

- Re-run
  `autolens_workspace_developer/jax_profiling/point_source/source_plane.py`.
- Tighten the full-pipeline JIT regression assertion (currently `rtol=2e-3`
  pending this work — see the `if full_pipeline_jits:` block in
  `source_plane.py` around line 522) back to `rtol=1e-4` against
  `EXPECTED_LOG_LIKELIHOOD_SOURCE_PLANE = -4496.798984131583`.
- Remove the "pending the np/jnp parity fix" comment pointing at this
  prompt.

## Affected repos

- `PyAutoGalaxy` (primary — `LensCalc.hessian_from` machinery, unit tests)
- `autolens_workspace_developer` (tighten the regression assertion in
  `point_source/source_plane.py` once the library fix lands)

No changes needed in `PyAutoLens` — the tracer layer calls down into the
same `LensCalc.hessian_from` method and inherits the fix automatically.

## Suggested branch

`feature/lens-calc-magnification-xp-parity`

## How to reproduce

```bash
source ~/Code/PyAutoLabs-wt/<task>/activate.sh   # or main checkout
cd /home/jammy/Code/PyAutoLabs/autolens_workspace_developer
python jax_profiling/point_source/source_plane.py
```

Look for the final assertion:

```
point_source/source_plane: regression — JIT log_likelihood drifted
 ACTUAL: array(-4491.832094)
 DESIRED: array(-4496.798984)
```

The drift will disappear once the magnification calc is aligned across
backends.

## Original analysis

Comparison script (standalone) that pins the divergence to magnifications:

```python
fit_np  = analysis_eager.fit_from(instance=instance)
fit_jax = analysis_jax.fit_from(instance=params_tree)

# residuals — identical
np.testing.assert_allclose(
    fit_np.positions.residual_map.array,
    np.asarray(fit_jax.positions.residual_map.array),
    rtol=1e-8,
)

# magnifications — diverge at ~5e-4 rel
print(fit_np.positions.magnifications_at_positions.array)
print(fit_jax.positions.magnifications_at_positions.array)
```

</details>

**Comment by Jammy2211** [comment]
## Library PR Created — 2026-04-18

### PR
- https://github.com/PyAutoLabs/PyAutoGalaxy/pull/358

### Status
Workspace changes needed. Next: `/start_workspace` on `autolens_workspace_developer`.

### API Changes
Internal numerics only — `LensCalc.hessian_from` numpy path now uses Richardson
extrapolation on the existing 2-point stencil (`(4·H(h/2) − H(h))/3`). Truncation
error improved from O(h²) ≈ 1e-4 to O(h⁴) ≈ 1e-8, matching the JAX path to
float64 precision. Public signature unchanged.

### Workspace follow-up required
- `jax_profiling/point_source/source_plane.py` — `EXPECTED_LOG_LIKELIHOOD_SOURCE_PLANE`
  was the stale O(h²) value (`-4496.798984`); new converged truth is ~`-4491.832205`
  (now agreed between np and jnp). Also tighten the `if full_pipeline_jits:`
  regression assertion from `rtol=2e-3` back to `rtol=1e-4` and drop the
  "pending parity fix" comment.
- `jax_profiling/point_source/image_plane.py` — re-run; update
  `EXPECTED_LOG_LIKELIHOOD_IMAGE_PLANE` if point-source likelihood shifted.
- `jax_profiling/imaging/*.py`, `jax_profiling/interferometer/*.py` — sanity
  re-run; unlikely to move since standard pipelines use analytic profile
  convergence rather than Hessian-derived quantities.

**Comment by Jammy2211** [comment]
## Shipped — 2026-04-18

### PRs
- https://github.com/PyAutoLabs/autolens_workspace_developer/pull/24

### Upstream PR
https://github.com/PyAutoLabs/PyAutoGalaxy/pull/358

### Summary
`LensCalc.hessian_from` numpy path now uses Richardson extrapolation on a 2-point central finite-difference stencil (`H = (4·H(h/2) − H(h))/3`), giving O(h⁴) accuracy and float64 parity with the JAX `jax.jacfwd` path. Closes the prior ~0.1% np/jnp log-likelihood divergence in source-plane point-source profiling. The `source_plane.py` regression assertion was updated to the new converged truth (`-4491.83220547254`) and JIT rtol tightened from `2e-3` to `1e-4`.

### Session Notes
- Dropped the originally-suggested option (a) — analytic Hessians via `convergence_2d_from` + `shear_yx_2d_from` — because it would have required every mass profile to expose those methods consistently, complicating the API. Kept as a future-work note in case a profile-by-profile rollout becomes worthwhile later.
- Dropped the 5-point central stencil too: same per-iteration cost as Richardson but more invasive (new code path) and not noticeably more accurate at this scale.
- Magnification parity test had to relax from `rtol=1e-8` (Hessian-component parity) to `rtol=1e-7`. The extra bit of float64 round-off comes from the `(1-hxx)(1-hyy) - hxy*hyx` → `1/det_A` chain — still ~4 orders of magnitude tighter than the 5e-4 divergence it replaces.
- Image-plane point-source likelihood is unaffected (uses only deflection angles, not Hessian-derived magnifications).

**Comment by Jammy2211** [comment]
Shipped via #358. Closing.
